simapp/sim_test.go

```diff
@@ -45,120 +44,186 @@ func init() {
+}
 
-	if config.Commit {
-		simtestutil.PrintStats(db)
+func TestAppStateDeterminism(t *testing.T) {
+	const numTimesToRunPerSeed = 3
+	var seeds []int64
+	if s := simcli.NewConfigFromFlags().Seed; s != simcli.DefaultSeedValue {
+		// We will be overriding the random seed and just run a single simulation on the provided seed value
+		for j := 0; j < numTimesToRunPerSeed; j++ { // multiple rounds
+			seeds = append(seeds, s)
+		}
+	} else {
+		// setup with 3 random seeds
+		for i := 0; i < 3; i++ {
+			seed := rand.Int63()
+			for j := 0; j < numTimesToRunPerSeed; j++ { // multiple rounds
+				seeds = append(seeds, seed)
+			}
+		}
 	}
-
-	t.Log("exporting genesis...\n")
-
-	exported, err := app.ExportAppStateAndValidators(false, []string{}, []string{})
-	require.NoError(t, err)
-
-	t.Log("importing genesis...\n")
-
-	newDB, newDir, _, _, err := simtestutil.SetupSimulation(config, "leveldb-app-sim-2", "Simulation-2", simcli.FlagVerboseValue, simcli.FlagEnabledValue)
-	require.NoError(t, err, "simulation setup failed")
-
-	defer func() {
-		require.NoError(t, newDB.Close())
-		require.NoError(t, os.RemoveAll(newDir))
-	}()
-
-	appOptions[flags.FlagHome] = newDir // ensure a unique folder for the new app
-	newApp := NewSimApp(log.NewNopLogger(), newDB, nil, true, appOptions, fauxMerkleModeOpt, baseapp.SetChainID(SimAppChainID))
-	require.Equal(t, "SimApp", newApp.Name())
-
-	var genesisState GenesisState
-	err = json.Unmarshal(exported.AppState, &genesisState)
-	require.NoError(t, err)
-
-	ctxA := app.NewContextLegacy(true, cmtproto.Header{Height: app.LastBlockHeight()})
-	ctxB := newApp.NewContextLegacy(true, cmtproto.Header{Height: app.LastBlockHeight()})
-	_, err = newApp.ModuleManager.InitGenesis(ctxB, genesisState)
-	if err != nil {
-		if strings.Contains(err.Error(), "validator set is empty after InitGenesis") {
-			t.Log("Skipping simulation as all validators have been unbonded")
-			t.Logf("err: %s stacktrace: %s\n", err, string(debug.Stack()))
-			return
+	// overwrite default app config
+	interBlockCachingAppFactory := func(logger log.Logger, db dbm.DB, traceStore io.Writer, loadLatest bool, appOpts servertypes.AppOptions, baseAppOptions ...func(*baseapp.BaseApp)) *SimApp {
+		if FlagEnableStreamingValue {
+			m := map[string]any{
+				"streaming.abci.keys":             []string{"*"},
+				"streaming.abci.plugin":           "abci_v1",
+				"streaming.abci.stop-node-on-err": true,
+			}
+			others := appOpts
+			appOpts = sims.AppOptionsFn(func(k string) any {
+				if v, ok := m[k]; ok {
+					return v
+				}
+				return others.Get(k)
+			})
 		}
+		return NewSimApp(logger, db, nil, true, appOpts, append(baseAppOptions, interBlockCacheOpt())...)
 	}
+	var mx sync.Mutex
+	appHashResults := make(map[int64][][]byte)
+	appSimLogger := make(map[int64][]simulation.LogWriter)
+	captureAndCheckHash := func(t *testing.T, ti sims.TestInstance[*SimApp]) {
+		seed, appHash := ti.Cfg.Seed, ti.App.LastCommitID().Hash
+		mx.Lock()
```

Comment:
Should the lock be set before getting the seed and appHash since are shared resources among the test instances?

Reply:
The TestInstance object and their data is not shared between instances. They work on a cloned object